cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java

```diff
@@ -720,6 +737,30 @@ public boolean isUpgradeRequired(MetadataScope scope) {
     return false;
   }
 
+  /**
+   * Scan the metadata store with a query which matches the index generated by
+   * {@link NearestKnownAncestorIndexer} to list all the sub entities which are associated with the given
+   * metadataEntity if it is a custom {@link MetadataEntity}.
+   * @param metadataEntity the {@link MetadataEntity} whose sub entities needs to be found
+   * @return a {@link Set} of {@link MetadataSearchResultRecordV2} which contains all the sub entities for the given
+   * metadataEntity if it is a custom entity else an empty set.
+   */
+  private Set<MetadataSearchResultRecordV2> getSubEntities(MetadataEntity metadataEntity) {
+    Set<MetadataSearchResultRecordV2> subEntities = new HashSet<>();
+    EntityId selfOrParentEntityId = EntityId.getNearestKnownEntity(metadataEntity);
+    // if a known entity is being deleted then we need to delete all it's sub-entity metadata too
+    if (selfOrParentEntityId.getEntityType().toString().equalsIgnoreCase(metadataEntity.getType())) {
+      // non-custom known type search for all subEntities which has been tagged with this entity as parent
+      SearchRequest request = new SearchRequest(new NamespaceId(metadataEntity.getValue(MetadataEntity.NAMESPACE)),
+                                                NearestKnownAncestorIndexer.PARENT_KEY +
```

Comment:
the fact that these implementation details need to be visible to this class indicates that this logic is in the wrong place. Does it make more sense to move this down into the dataset?

Reply:
Initially I did put it there but moved it out so that search and other deletes happen in their own independent transactions as having it in dataset will make them run under one transaction which has high possibility of timing out. 
Agree that the current way is not ideal too as it will lead to large number of transaction or even a http timeout in worse case. Ideally, we should be doing this in batch operation across multiple transaction but that calls for a much bigger refactoring which we have tried to avoid.

For the logic being here I am guessing you are referring to how we are generating the search query. Even moving it to the dataset will have similar issue as the index is generated by NearestKnownEntityIndexer. I can make it expose something to encapsulate this.